contracts/strategies/aura/base/AuraStrategyAuraBAL.sol

```diff
@@ -0,0 +1,710 @@
+      swapPoolId, 
+      IBVault.SwapKind.GIVEN_IN,
+      IAsset(universalRewardToken),
+      IAsset(underlyingToken),
+      tokenBalance, 
+      bytes(""));
+    
+    // Swap with the limit that we need to receive more than given in
+    IBVault(bVault).swap(
+      swapInfos,
+      funds,
+      tokenBalance,
+      999999999999999999
+    );
+    uint256 underlyingAfter = IERC20(underlyingToken).balanceOf(address(this));
+
+    // If the underlying balance after swapping increased,
+    // The swap succeeds and the strategy received more from swapping
+    if (underlyingAfter <= underlyingBefore) {
+      // we can accept 0 as minimum, this will be called only by trusted roles
+      uint256 depositTokensAmount = depositNTokens();
+      IAsset[] memory depositAssets = new IAsset[](depositTokensAmount);
+      for (uint256 i = 0; i < depositTokensAmount; i++) {
+        depositAssets[i] = IAsset(poolAssets[i]);
+      }
+
+      IBVault.JoinKind joinKind = IBVault.JoinKind.EXACT_TOKENS_IN_FOR_BPT_OUT;
+      uint256[] memory amountsIn = new uint256[](depositTokensAmount);
+      amountsIn[depositArrayIndex()] = tokenBalance;
+      uint256 minAmountOut = 1;
+
+      bytes memory userData = abi.encode(joinKind, amountsIn, minAmountOut);
+
+      IBVault.JoinPoolRequest memory request;
+      request.assets = depositAssets;
+      request.maxAmountsIn = amountsIn;
+      request.userData = userData;
+      request.fromInternalBalance = false;
+
+      IBVault(bVault).joinPool(
+        balancerDepositPoolId(),
+        address(this),
+        address(this),
+        request
+      );
+
+      address depositor = auraDeposit();
+      address universalDepositToken = depositToken();
+      uint256 bptBalance = IERC20(universalDepositToken).balanceOf(address(this));
+      IERC20(universalDepositToken).safeApprove(depositor, 0);
+      IERC20(universalDepositToken).safeApprove(depositor, bptBalance);
+      IAuraDepositor(depositor).deposit(bptBalance, true); //deposit and stake
+    }
+  }
+
+  /**   Stakes everything the strategy holds into the reward pool
+   */
+  function _investAllUnderlying() 
+    internal 
+    onlyNotPausedInvesting 
+  {
+    // this check is needed, because most of the SNX reward pools will revert if
+    // you try to stake(0).
+    if(IERC20(underlying()).balanceOf(address(this)) > 0) {
+      _enterRewardPool();
+    }
+  }
+
+  // Reward => Deposit(WETH) => JoinPool Bpt => Deposit(Bpt) => StakeAll
+  function _enterRewardPool() 
+    internal 
+  {
+    address lpToken = underlying();
+    address stakePool = rewardPool();
+    uint256 entireBalance = IERC20(lpToken).balanceOf(address(this));
+    IERC20(lpToken).safeApprove(stakePool, 0);
+    IERC20(lpToken).safeApprove(stakePool, entireBalance);
+    IAuraBaseRewardPool(stakePool).stakeAll(); // Stake
+  }
+
+  function addRewardToken(
+    address _token, 
+    address[] memory _path2Reward, 
+    bytes32 _dexOption
```

Comment:
The dexes could be inputed as a list too, to generalise for tokens that would need to be liquidated using multiple dexes. This is the implementation I made for an updated version of the Convex strategy: https://github.com/harvest-finance/harvest-strategy/pull/77/files#diff-e91096217bcc49a83d27481a3ff62b6f23f59ab99b9ebae1edf466798d38754eR192